internal/clients/clients.go

```diff
@@ -0,0 +1,108 @@
+package clients
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/selection"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/tools/clientcmd"
+	clusterv1 "open-cluster-management.io/api/cluster/v1"
+	clusterv1beta1 "open-cluster-management.io/api/cluster/v1beta1"
+	clusterv1beta2 "open-cluster-management.io/api/cluster/v1beta2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+
+	monitoringv1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+)
+
+const (
+	ListRequestTimeout   = 30 * time.Second
+	SingleRequestTimeout = 10 * time.Second
+
+	secretTypeLabel      = "hive.openshift.io/secret-type"
+	secretTypeLabelValue = "kubeconfig"
+)
+
+// NewClientForHub creates a new client for the hub cluster
+func NewClientForHub() (client.Client, error) {
+	conf, err := config.GetConfig()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get config: %w", err)
+	}
+
+	c, err := client.New(conf, client.Options{Scheme: GetSchemeForHub()})
+	if err != nil {
+		return nil, fmt.Errorf("failed to create client: %w", err)
+	}
+
+	return c, nil
+}
+
+// GetSchemeForHub returns the scheme for the hub cluster client
+func GetSchemeForHub() *runtime.Scheme {
+	scheme := runtime.NewScheme()
+	utilruntime.Must(corev1.AddToScheme(scheme))
+	utilruntime.Must(clusterv1.AddToScheme(scheme))
+	utilruntime.Must(clusterv1beta1.AddToScheme(scheme))
+	utilruntime.Must(clusterv1beta2.AddToScheme(scheme))
+
+	return scheme
+}
+
+// NewClientForCluster creates a new client for a managed cluster
+func NewClientForCluster(ctx context.Context, hubClient client.Client, clusterName string) (client.Client, error) {
+	kubeConfig, err := getClusterKubeConfigFromSecret(ctx, hubClient, clusterName)
+	if err != nil {
+		return nil, err
+	}
+
+	conf, err := clientcmd.RESTConfigFromKubeConfig(kubeConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get config: %w", err)
+	}
+
+	c, err := client.New(conf, client.Options{Scheme: GetSchemeForCluster()})
```

Comment:
Instead of creating many clients for many kubeconfigs...is it possible to have multiple kubeconfing merged into one and thne switch context as needed? Client i think is also concurrency safe. https://pkg.go.dev/k8s.io/client-go@v0.26.3/tools/clientcmd

Reply:
I don't really see the benefit of doing that. Besides, we won´t be accessing the clusters often. Also, we would need to cover scenarios such as the deletion of a managed cluster. And I'm using concurrency to get the rules. This would mean to stop to merge the configs and then continue.

This is simple and not expensive. It creates a client (not expensive in my opinion), get the stuff and goes out of scope so it can be collected by the GC.

Reply:
sounds good